### PR TITLE
Add createCustomColorsHOC HOC builder

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
This PR is useful for #10611.
Supersedes: https://github.com/WordPress/gutenberg/pull/10765

createCustomColorsHOC builds a HOC equivalent to withColors but that has a static color array instead of using the color array provided by the theme or the default colors.

Props to @talldan for suggesting this feature/interface.

Sample:
```
 var BORDER_COLORS = [ {
            slug: 'green',
            name: 'Green',
            color: '#00ff00'
        },{
            slug: 'yellow',
            name: 'Yellow',
            color: '#ffff00'
    } ];
var withBorderColors = createCustomColorsHOC( BORDER_COLORS );
// ...
compose( [ 
	withColors( 'backgroundColor', 'textColor' ),
	withBorderColors( 'borderBottomColor', 'borderTopColor' )
] );
```
In the previous sample backgroundColor and textColor use the color palette provided by the editor/theme while borderBottomColor and borderTopColor use the static colors provided by the block.
The block is responsible for implementing the following classes: .has-green-border-top-color, has-yellow-border-top-color, has-green-border-bottom-color, .has-yellow-border-bottom-color;
This is useful in a context where there are no expectations themes should provide classes e.g: border colors, or if the block only makes sense with a given set of colors and we want to avoid inline styles.

## How has this been tested?
I added the test bock available in https://gist.github.com/jorgefilipecosta/74ba266ed16a6759c85e1a13effe447c.

I checked that for border colors we can choose between a specified set, and classes are used for border colors with the logic being handled by the component created by createCustomColorsHOC (equivalent to) withColors.
I checked that background and text color still present the expected behavior.
I checked that existing blocks that use colors ( cover, paragraph, button etc...) still work as expected.

### CODE REVIEW

**Review by talldan — COMMENTED** [review]
Hi @jorgefilipecosta Thanks for working on this a while back! Would be good to get the ball rolling on this again now that 5.0 is out so that I can unblock the task for adding background colors to the table block.

I hope you don't mind but I've added a couple of commits to resolve a couple of things I spotted when reviewing - firstly that createCustomColorsHOC doesn't need to use withSelect. I've refactored so that it doesn't. Secondly that the display name wasn't quite right, so I've resolved that.

I've also added some tests for createCustomColorsHOC which I think is useful as it also adds some coverage for the code it shares with withColors.

File: packages/editor/src/components/colors/with-colors.js [hunk 1]
```diff
@@ -43,6 +46,11 @@ export default ( ...args ) => {
 	return createHigherOrderComponent(
 		compose( [
 			withSelect( ( select ) => {
+				if ( colorsArray ) {
```
[the remark this answers is not in the document]
Comment (talldan) [reply packages/editor/src/components/colors/with-colors.js]: Fixed in bfbb128

File: packages/editor/src/components/colors/with-colors.js [hunk 2]
```diff
@@ -132,3 +140,5 @@ export default ( ...args ) => {
 		'withColors'
```
[the remark this answers is not in the document]
Comment (talldan) [reply packages/editor/src/components/colors/with-colors.js]: Fixed in bfbb128

**Comment by talldan** [comment]
Closing this one as the change was merged in #10611

**Comment by jorgefilipecosta** [comment]
Thank you for taking these changes to the end @talldan!

**Pull request closed without merging**